[Java] Added method to remove null_masks if the column has no nulls

razajafri opened the pull request:
This PR adds a method to remove the validity vector in cases where there are columns in a Table with no nulls but still have a validity vector. 


<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

revans2 commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,6 +935,39 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
+  if (column_view.nullable() && column_view.null_count() == 0) {
+    // null_mask is allocated but no nulls present therefore we create a new column_view without the
+    // null_mask to avoid things blowing up in reading the parquet file
+    if (!cudf::is_nested(column_view.type())) {
```
A child column may still need to be modified, even if a parent column is not nullable.

davidwendt commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,6 +935,39 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
+    if (!cudf::is_nested(column_view.type())) {
+      cudf::column_view cv(column_view.type(), column_view.size(), column_view.data<int32_t>(),
+                           nullptr, 0);
+      return cv;
```
It appears this will only work with INT32 columns. 
Also, sliced column's will have an offset which should be passed on the ctor as well.

ttnghia replied:
Like David said, if you want to support generic type, use `void` type.
```suggestion
      return cudf::column_view(column_view.type(), column_view.size(), column_view.template data<void>(), nullptr, 0, column_view.offset());
```

razajafri replied:
I am now using `head()` which is better I think because we don't have to worry about `offset`

davidwendt commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,6 +935,39 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
+      for (auto it = column_view.child_begin(); it != column_view.child_end(); it++) {
+        children.push_back(remove_validity_from_col(*it));
+      }
+      cudf::column_view cv(column_view.type(), column_view.size(), column_view.data<int32_t>(),
+                           nullptr, 0, 0, children);
```
```suggestion
      cudf::column_view cv(column_view.type(), column_view.size(), nullptr,
                           nullptr, 0, column_view.offset(), children);
```
Parent columns have no data and the `data()` method accounts for `offset` so it would be bad to pass what it returns here.

revans2 commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,6 +935,39 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
+  }
+}
+
+cudf::table_view remove_validity_if_needed(cudf::table_view *input_table_view) {
```
As a side note, it might be good to try and put this into the C++ part of cudf as a stand alone API.  From your tests Spark has trouble reading maps from parquet where the key is nullable even if there are no null values. That means anyone trying to write out maps in parquet through cudf is going to be producing files that are not compatible with Spark. If these are official cudf APIs, then they have the option to fix it., even if it is just for the keys to maps.

davidwendt replied:
Since each call to `null_count()` could call a kernel, ideally this would be more efficient as a single-kernel across all bitmasks in a table: Reference #8365

revans2 replied:
That would also be a great API to have.  One that can look at all of the columns in a table recursively, see which already have `null_count`s cached, and then calculate/cache them for any that are missing.

razajafri replied:
In interest of time, I would love to do #8365 as a follow-on if you are OK with it.

revans2 replied:
I am fine with it as a follow on.

ttnghia commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,6 +935,39 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
   return combine_join_results(env, std::move(left_cols), std::move(right_cols));
 }
 
+cudf::column_view remove_validity_from_col(cudf::column_view column_view) {
```
```suggestion
cudf::column_view remove_validity_from_col(cudf::column_view const& column_view) {
```

ttnghia commented on java/src/main/java/ai/rapids/cudf/ColumnView.java:
```diff
@@ -140,6 +140,23 @@ public final DType getType() {
+   * Please note that it is the responsibility of the caller to close these views.
+   * @return an array of child column views
+   */
+  public final ColumnView[] getChildColumnViews() {
+    int numChildren = getNumChildren();
+    if (!getType().isNestedType()) {
+      return null;
+    }
+    ColumnView[] views = new ColumnView[numChildren];
+    for (int i = 0 ; i < numChildren ; i++) {
+      views[i] = getChildColumnView(i);
+    }
+    return views;
```
This seems wrong to me. The results is an array of column views of nulls?

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6075,68 @@ void testAllFilteredFromValidity() {
+                   new DeviceMemoryBuffer(nonNullVector1.getOffsets().address,
+                       nonNullVector1.getOffsets().length,
+                       Cuda.DEFAULT_STREAM), null,
+               Arrays.stream(nonNullVector1.getChildColumnViews()).mapToLong((cv)-> cv.getNativeView()).toArray());
```
You are not closing the child column views.  This code is leaking them.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6075,68 @@ void testAllFilteredFromValidity() {
+             Arrays.asList(structData1, structData2),
+             Arrays.asList(structData1, structData2))) {
+      //Then we take the created ColumnVectors and add validity masks even though the nullCount = 0
+      try (ColumnVector cv0 = new ColumnVector(nonNullVector0.type, nonNullVector0.rows,
```
I personally would prefer to see a stand along method to add in the validity to a ColumnView.

```
ColumnView replaceValidity(ColumnView original, DeviceMemoryBuffer validity) {
    assert(validity.length >= BitVectorHelper.getValidityAllocationSizeInBytes(original.rows));
    // get children
    try {
        return new ColumnView(original.type,
            original.rows,
            Optional<Long>.empty(),
            validity,
            original.getOffsets(),
            children);
    } finally {
          // close children
    }   
}
```

Then we can insert this into children columns if we want.  We can copy it back out to a ColumnVector if we need to.  It is just a lot more flexible and cleaner.

razajafri replied:
Do you have concerns with making this a instance method on ColumnView?

razajafri replied:
nm, I misunderstood

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6075,68 @@ void testAllFilteredFromValidity() {
+
+        try (Table t = new Table(new ColumnVector[]{cv0, cv1, cv2});
+             Table tableWithoutNullMask = removeNullMasksIfNeeded(t)) {
+          for (int i = 0; i < tableWithoutNullMask.getNumberOfColumns(); i++) {
```
This is only testing the parent columns. It is not testing the child columns.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6075,68 @@ void testAllFilteredFromValidity() {
+           ColumnVector cv1 = new ColumnVector(nonNullVector1.type, nonNullVector1.rows,
+               Optional.of(0L),
+               null,
+               DeviceMemoryBuffer.allocate(BitVectorHelper.getValidityLengthInBytes(nonNullVector1.rows)),
```
This is uninitialized memory. We probably want to do a mem-set on it so there is no question what is in it.

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -1121,7 +1121,7 @@ public static ColumnVector boolFromBytes(byte... values) {
 
   /**
-   * This method is evolving, unstable and currently test only.
+   * This method is evolving and unstable.
```
Someone made them public, I just modified the comment to match the method

revans2 replied:
Even though they are public they still are intended to only be used for tests.  The comment was accurate

davidwendt commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,13 +935,77 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
 
+cudf::column_view remove_validity_from_col(cudf::column_view column_view) {
+  std::unique_ptr<cudf::column_view> ret;
+  if (!cudf::is_nested(column_view.type())) {
```
```suggestion
  if (!cudf::is_compound(column_view.type())) {
```
The `is_compound()` will be true for all the nested types as well as string (and dictionary). So you don't have special code for string here.

razajafri commented:
Thanks for the review guys, is there anything else that you want me to change?

davidwendt commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -935,13 +935,70 @@ jlongArray combine_join_results(JNIEnv *env, cudf::table &left_results,
+    for (auto it = column_view.child_begin(); it != column_view.child_end(); it++) {
+      children.push_back(remove_validity_from_col(*it));
+    }
+    if (!column_view.nullable() || !column_view.null_count() == 0) {
+      ret.reset(new cudf::column_view(column_view.type(), column_view.size(), nullptr,
+                                      column_view.null_mask(), 0, column_view.offset(), children));
```
Looks like you are missing the null count here?
```suggestion
    if (!column_view.nullable() || column_view.null_count() != 0) {
      ret.reset(new cudf::column_view(column_view.type(), column_view.size(), nullptr,
                                      column_view.null_mask(), column_view.null_count(), column_view.offset(), children));
```

or maybe
```suggestion
    if (!column_view.nullable() || column_view.null_count() != 0) {
      ret.reset(new cudf::column_view(column_view));
```

razajafri replied:
Yes, good catch

razajafri commented:
Rerun tests

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnView.java:
```diff
@@ -91,6 +91,7 @@ public ColumnView(DType type, long rows, Optional<Long> nullCount,
    * @param validityBuffer an optional validity buffer. Must be provided if nullCount != 0.
    *                       The ownership doesn't change on this buffer
    */
+  @Deprecated
```
Why is this deprecated?  You can create most column views in a perfectly valid way with this?  Just not nested ones.

razajafri replied:
I deprecated this because 1. It doesn't take in an offset buffer which is needed to create Strings and 2. Because ColumnVector doesn't have a similar API. But I do agree this is a much simpler API and makes more sense. We should change the way we expect the users of Java API to think of Strings as columns with no children.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnView.java:
```diff
@@ -101,11 +102,39 @@ public ColumnView(DType type, long rows, Optional<Long> nullCount,
+   *                       The ownership doesn't change on this buffer
+   * @param offsetBuffer   The offsetbuffer for columns that need an offset buffer
+   */
+  public ColumnView(DType type, long rows, Optional<Long> nullCount,
```
This API really confuses me, because why would we need to create a ColumnView that has offsets in it, but no child?  In List columns there is always a child.  We hacked it up badly for Strings in the java side by putting them together, but we should not have done that.

razajafri replied:
This API mirrors ColumnVector's API. Strings column is a hybrid column on the Java side as we make it seem like it doesn't have children but has an offset buffer and a data buffer where on the cudf side it the parent only has a validity buffer, the data and offsets are children.

Ideally we should create a String column like we create a List, then we won't need a special case just for Strings, right now String is stuck in between the `nested` and `non-nested` type in Java. 

I don't know how else to do this other than changing how we are creating Strings

revans2 replied:
Sorry I am not asking you to change how we do Strings. I am just trying to understand why these changes are being made, and why one API is deprecated in favor of this one. I see now that it follows the public API in ColumnVector and is a superset of the old API. I am still not convinced we need to deprecate the old API, but this is fine.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnView.java:
```diff
@@ -140,6 +169,32 @@ public final DType getType() {
+        views[i] = getChildColumnView(i);
+      }
+      return views;
+    } catch(Exception e) {
```
I would prefer to can Throwable, as this is likely missing OOM errors.

razajafri commented:
@revans2 I have addressed your concerns. I have removed deprecation as it doesn't hurt to keep this API it also makes more sense and is less confusing to users.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6077,167 @@ void testAllFilteredFromValidity() {
+             Arrays.asList(structData1, structData2))) {
+      //Then we take the created ColumnVectors and add validity masks even though the nullCount = 0
+      long allocSize = BitVectorHelper.getValidityAllocationSizeInBytes(nonNullVector0.rows);
+      DeviceMemoryBuffer dm0 = DeviceMemoryBuffer.allocate(allocSize);
```
This buffer is leaked.

razajafri replied:
sorry, I assumed the ColumnVector will take ownership of the buffer and close it when its closed, not thinking that we are creating a ColumnView and copying it to a ColumnVector

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -6080,6 +6077,167 @@ void testAllFilteredFromValidity() {
+           ColumnView newStructChild0 = replaceValidity(structChild0, dm3_child);
+           ColumnView newStruct = struct.replaceChildrenWithViews(new int[]{0}, new ColumnView[]{newStructChild0});
+           ColumnView list = nonNullVector1.replaceChildrenWithViews(new int[]{0}, new ColumnView[]{newStruct});
+           ColumnVector cv1 = replaceValidity(list, dm1).copyToColumnVector();
```
The view from the replace validity is leaked.

razajafri replied:
I dislike the name `getChildColumnView` considering that we have a `getColumn` which simply returns a reference to an existing vector and no cleanup is necessary. I wish we could name it something like `createChildColumnView` so the user knows that a new `ColumnView` is being returned and needs to be cleaned. But tbh I don't know if I am the only one who seems to miss cleaning up the ColumnViews or if there are others.

razajafri commented:
Can I ask for one more review?

razajafri commented:
rerun tests

dillon-cullinan commented:
rerun tests

razajafri commented:
@gpucibot merge
